### PR TITLE
Loop Conversion

### PR DESCRIPTION
**Pull request by abhi-iyer**
# Description

Dynamic loop conversion from PyTorch to TRT. A few limitations with this approach, as of now:

- all recurrent values inside loop can be Tensors only
- works with for loops, albeit with a comparative threshold (i.e. TRT vs JIT) that is greater than the set limit of 2e-5
- breaks converter contract: all primitive types (e.g. int, float, etc.) could potentially be Tensors when processed inside the converter

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation and have regenerated the documentation (`make html` in docsrc)
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by abhi-iyer** [comment]
For more information on Loop Conversion, please see:

[TensorRT Loops - A Rundown.pdf](https://github.com/NVIDIA/TRTorch/files/4991221/TensorRT.Loops.-.A.Rundown.pdf)

**Comment by abhi-iyer** [comment]
Currently in the testing phase. Please test with the following code snippet first (note: this is not exactly an RNN cell, but it's okay to test with this for now):

```
import torch.nn as nn
import torch
import torch.nn.functional as F
import trtorch

trtorch.logging.set_reportable_log_level(trtorch.logging.Level.Debug)

class RNN(nn.Module):
  def __init__(self, W_h, U_h, W_y, b_h, b_y):
    super(RNN, self).__init__()
    self.W_h = nn.Parameter(W_h)
    self.U_h = nn.Parameter(U_h)
    self.W_y = nn.Parameter(W_y)
    self.b_h = nn.Parameter(b_h)
    self.b_y = nn.Parameter(b_y)

  def forward(self, x, h):
    y = torch.tanh(h)

    for t in range(x.size(0)):
      h = torch.tanh(x[t] @ self.W_h + h @ self.U_h + self.b_h)
      y = torch.tanh(h @ self.W_y + self.b_y)

    return y, h

x = torch.randn(5, 4, 3).to('cuda')
h = torch.randn(4, 4).to('cuda')

rnn = RNN(torch.randn(3, 4), torch.randn(4, 4), torch.randn(4, 4), torch.randn(4, 4), torch.randn(4, 4)).eval().cuda()
script = torch.jit.script(rnn)

print(script.graph)

torch.jit.save(script, '<path>/script.ts')
```

File: core/conversion/conversion.cpp [hunk 1]
```diff
@@ -291,8 +298,141 @@ void EvaluateLoopBlock(ConversionCtx* ctx, const torch::jit::Node* n) {
...
+    auto max_trip_count = ctx->evaluated_value_map[n->input(0)];
+    auto start_cond = ctx->evaluated_value_map[n->input(1)];
+
+    ctx->evaluated_value_map[block->inputs()[0]] = torch::jit::IValue(0);
+    auto trip_count = ctx->evaluated_value_map[block->inputs()[0]];
+
+    // map node inputs [recurrent values] -> node outputs [recurrent values]
+    MapIValues(ctx, n->inputs(), n->outputs(), 2, 0);
+
+    LOG_DEBUG(ctx->logger, "(Loop Evaluation) Evaluating loop " << *n);
```
Comment (narendasan) [inline core/conversion/conversion.cpp:314]: Can you change the prefix here to Loop Conversion so we can tell the difference?

File: core/conversion/conversion.cpp [hunk 2]
```diff
@@ -291,8 +298,141 @@ void EvaluateLoopBlock(ConversionCtx* ctx, const torch::jit::Node* n) {
...
+        auto inp = n->inputs()[i];
+
+        if (inp->type()->isSubtypeOf(c10::TensorType::get())) {
+            auto recur = loop->addRecurrence(*ctx->value_tensor_map[inp]);
+            TRTORCH_CHECK(recur, "Unable to create recurrent layer from node: " << *n);
+
+            std::ostringstream tensor_id;
+            tensor_id << reinterpret_cast<int*>(ctx->value_tensor_map[inp]);
+
+            recur->setName(("recurrent layer for Tensor " + tensor_id.str()).c_str());
```
Comment (narendasan) [inline core/conversion/conversion.cpp:372]: We could just set the name to `{Value.name} [Recurrence Layer]`

File: core/conversion/conversion.cpp [hunk 3]
```diff
@@ -291,8 +298,141 @@ void EvaluateLoopBlock(ConversionCtx* ctx, const torch::jit::Node* n) {
...
+        } else {
+            TRTORCH_THROW_ERROR("Only recurrent Tensors allowed as input to Loop");
+        }
+    }
+
+    // evaluate/convert all nodes inside block
+    for (auto bn : block->nodes()) {
+        if (bn->kind() == torch::jit::prim::If) {
+            EvaluateConditionalBlock(ctx, bn, true);
+        } else if (evaluators::shouldEvalAtConversionTime(bn)) {
```
Comment (narendasan) [inline core/conversion/conversion.cpp:383]: Is there a reason we couldnt add support for nested loops both evaluated and converted  trivially here?

Comment (abhi-iyer) [reply core/conversion/conversion.cpp:383]: "nested loops" - `prim::Loops` or `prim::If`? or both?

**Comment by abhi-iyer** [comment]
Closing this version. Opening a new PR for loop conversion.

**Pull request closed without merging**